### ISSUE TITLE
Misclassified C as C++ in data.table project

### ISSUE THREAD

**Issue by jangorecki** [github/linguist#5028]
This repository does not use any C++ code but some header files are classified as such:
https://github.com/Rdatatable/data.table
It is strange that h files are not considered as C unless C++ syntax is detected inside, I think that should be the default option.
I would appreciate fixing it.

**Comment by smola** [comment]
@jangorecki This is a known problem: https://github.com/github/linguist/issues/1626

You can workaround it by overriding the language of `*.h` files in the repository with `.gitattributes`: https://github.com/github/linguist#overrides

**Comment by jangorecki** [comment]
I think it should work other way around. *.h classified as C but actually being C++, those should require workarounds.

**Comment by smola** [comment]
@jangorecki I agree. But there's some technical limitations that need to be solved before that's possible. There's a lengthy discussion about it at #1626. Short-term, overriding with `.gitattributes` is the only solution.

**Comment by jangorecki** [comment]
Issue is still relevant and mentioned technical limitations doesn't seem to be addressed.

**Comment by Starwort** [comment]
Also occurring in my repository:
https://github.com/Starwort/NEA/search?l=c%2B%2B
This seems very common...

**Comment by jangorecki** [comment]
Yes, h files should be recognized as C till the proper fix will be implemented.

**Comment by smola** [comment]
> Also occurring in my repository:
> https://github.com/Starwort/NEA/search?l=c%2B%2B

Note that these examples are probably unrelated to this PR, since they don't seem to be related to the number of braces, but the length of the files. The currently classifier can perform particularly bad in very short files.

**Comment by 0hana** [comment]
I also have this issue
https://github.com/0hana/C/test.graph.h

**Comment by lildude** [comment]
I'm closing this issue as it's essentially a duplicate of https://github.com/github/linguist/issues/1626 which I have now pinned to try and reduce the duplicates.

**Comment by jangorecki** [comment]
@lildude so till the proper solution is in place maybe we can just change classification of *.h files to be C rather than C++?

**Comment by lildude** [comment]
It's not that simple. `.h` is already classified as C:

https://github.com/github/linguist/blob/8011321f1394b5c3fc1f8ef242775af640df8f7d/lib/linguist/languages.yml#L614-L620

However, it is also classified as C++:

https://github.com/github/linguist/blob/8011321f1394b5c3fc1f8ef242775af640df8f7d/lib/linguist/languages.yml#L646-L661

... and Objective-C:

https://github.com/github/linguist/blob/8011321f1394b5c3fc1f8ef242775af640df8f7d/lib/linguist/languages.yml#L3847-L3857

... because it's valid in all three. Simply stripping from all but C will cause even more misclassification than currently seen with the current heuristics and classifier.

Where things go wrong is when a file doesn't match any of the heuristics:

https://github.com/github/linguist/blob/aad49acc0624c70d654a8dce447887dbbc713c7a/lib/linguist/heuristics.yml#L226-L231

So it falls through to the classifier which then has to make a guess based on the content of the samples it has. This is where things go wrong most of the time, especially with very small files.

What we could do, which may be the best solution, is to completely remove the classifier step for `.h` files, which is effectively what https://github.com/github/linguist/issues/1626#issuecomment-818682874 is suggesting.

I've started a PR at https://github.com/github/linguist/pull/5357 to implement this.
